src/main/cli/cli.c

```diff
@@ -6121,120 +6235,176 @@ static void cliHelp(char *cmdline)
+                    break;
+                if (!pstart->name[bufferIndex] && bufferIndex < sizeof(cliBuffer) - 2) {
+                    /* Unambiguous -- append a space */
+                    cliBuffer[bufferIndex++] = ' ';
+                    cliBuffer[bufferIndex] = '\0';
+                    break;
                 }
-                bufferIndex = 0;
+                cliBuffer[bufferIndex] = pstart->name[bufferIndex];
+            }
+        }
+        if (!bufferIndex || pstart != pend) {
+            /* Print list of ambiguous matches */
+            cliPrint("\r\033[K");
+            for (cmd = pstart; cmd <= pend; cmd++) {
+                cliPrint(cmd->name);
+                cliWrite('\t');
             }
+            cliPrompt();
+            i = 0;    /* Redraw prompt */
+        }
+        for (; i < bufferIndex; i++)
+            cliWrite(cliBuffer[i]);
+    } else if (!bufferIndex && c == 4) {   // CTRL-D
+        cliExit(cliBuffer);
+        return;
+    } else if (c == 12) {                  // NewPage / CTRL-L
+        // clear screen
+        cliPrint("\033[2J\033[1;1H");
+        cliPrompt();
+    } else if (c == 127) {
+        // backspace
+        if (bufferIndex) {
+            cliBuffer[--bufferIndex] = 0;
+            cliPrint("\010 \010");
+        }
+    } else {
+        processCommandCharacter(c);
+    }
+}
 
-            memset(cliBuffer, 0, sizeof(cliBuffer));
+void cliProcess(void)
+{
+    if (!cliWriter) {
+        return;
+    }
 
-            // 'exit' will reset this flag, so we don't need to print prompt again
-            if (!cliMode)
-                return;
+    // Flush the buffer to get rid of any MSP data polls sent by configurator after CLI was invoked
+    cliWriterFlush();
 
-            cliPrompt();
-        } else if (c == 127) {
-            // backspace
-            if (bufferIndex) {
-                cliBuffer[--bufferIndex] = 0;
-                cliPrint("\010 \010");
-            }
-        } else if (bufferIndex < sizeof(cliBuffer) && c >= 32 && c <= 126) {
-            if (!bufferIndex && c == ' ')
-                continue; // Ignore leading spaces
-            cliBuffer[bufferIndex++] = c;
-            cliWrite(c);
+    while (serialRxBytesWaiting(cliPort)) {
+        uint8_t c = serialRead(cliPort);
+
+        processCharacter(c);
+    }
+}
+
+#if defined(USE_CUSTOM_DEFAULTS)
+void cliProcessCustomDefaults(void)
+{
+    if (processingCustomDefaults) {
+        return;
+    }
+
+    char *ptr = customDefaultsStart;
+    if (isDefaults(ptr)) {
+#if !defined(DEBUG_CUSTOM_DEFAULTS)
+        bufWriter_t *cliWriterTemp = cliWriter;
+        cliWriter = NULL;
+#endif
+        memcpy(cliBufferTemp, cliBuffer, sizeof(cliBuffer));
+        uint32_t bufferIndexTemp = bufferIndex;
+        bufferIndex = 0;
+        processingCustomDefaults = true;
+
+        while (*ptr && ptr != 0xFF && ptr < customDefaultsEnd) {
+            processCommandCharacter(*ptr++);
```

Comment:
I think this should be 'processCharacter' here ;D

Reply:
Ahh, no you're right, I was mis-reading..

I think that the confusion was caused by the name of the methods.

I think I was making a bridge between the words 'command' and 'control', and then thinking that this was only processing 'command' characters (i.e. the non-printable ascii stuff).

Perhaps there are better names for `processCommandCharacter` and `processCharacter` methods?  I think the main difference here is interactivity.  i.e. we want to exclude interactive control characters when processing 'custom defaults'.  so perhaps: `processNonInteractiveCharacter` and `processAnyCharacter` ? I dunno, but I do no that 'command' caused me some confusion.

While we're talking about naming, perhaps `ptr`, which is very generic, could also be renamed to `defaultsBuffer`?

Reply:
As someone who fakes knowing C the pointer part is important. 
`defaultsBufferPointer`? `defaultsBufferPtr`

Reply:
By all means, this is not a buffer, as we are not buffering anything. But made it more specific.